alacritty_terminal/src/input.rs

```diff
@@ -408,21 +431,37 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
             return;
         }
 
-        // Only report motions when cell changed and mouse is not over the message bar
-        if let Some(message) = self.message_at_point(Some(point)) {
-            self.update_message_cursor(point, message);
-
-            return;
-        } else {
-            self.ctx.terminal_mut().reset_mouse_cursor();
-        }
-
         // Don't launch URLs if mouse has moved
         self.ctx.mouse_mut().block_url_launcher = true;
 
-        // Underline URLs and change cursor on hover
-        if cell_changed {
-            self.update_url_highlight(point, modifiers);
+        match self.mouse_position(point) {
+            MousePosition::Url(url) => {
+                let mouse_mode =
+                    TermMode::MOUSE_MOTION | TermMode::MOUSE_DRAG | TermMode::MOUSE_REPORT_CLICK;
+
+                if self.mouse_config.url.mods().relaxed_eq(modifiers)
+                    && (!self.ctx.terminal().mode().intersects(mouse_mode) || modifiers.shift)
+                    && self.mouse_config.url.launcher.is_some()
+                {
+                    let url_bounds = self.url_bounds_at_point(url, point);
+                    self.ctx.terminal_mut().set_url_highlight(url_bounds);
+                    self.ctx.terminal_mut().set_mouse_cursor(MouseCursor::Hand);
+                    self.ctx.terminal_mut().dirty = true;
+                }
+            },
+            MousePosition::MessageBar => {
+                self.ctx.terminal_mut().set_mouse_cursor(MouseCursor::Default);
+                return;
+            },
+            MousePosition::MessageBarButton => {
+                self.ctx.terminal_mut().set_mouse_cursor(MouseCursor::Hand);
+                return;
+            },
+            MousePosition::Terminal => {
+                // `reset_url_highlight` calls `reset_mouse_cursor`, so no need to perform explicit
+                // call to the second one.
+                self.ctx.terminal_mut().reset_url_highlight();
```

Comment:
I haven't tested this, but I'm a bit concerned about calling `reset_url_highlight` only when the mouse is over the terminal. What if the URL is in the last line and there is a message bar open? Wouldn't that cause the URL to stay highlighted until the mouse is moved to the terminal again?

I'd consider an `if/else` block here too where the `else` always resets url highlight, however I think that might get very deeply nested and convoluted. So I think I'd just go for putting `reset_url_highlight` in every block and making sure it doesn't reset the mouse cursor itself.

Reply:
Good one, I'll test this scenario and update.

> reset_url_highlight in every block and making sure it doesn't reset the mouse cursor itself.

After removing `reset_mouse_cursor` from `reset_url_highlight` everything should be very simple to handle.